src/lamp_py/bus_performance_manager/events_tm_schedule.py

```diff
@@ -101,85 +98,97 @@ def generate_tm_schedule() -> TransitMasterTables:
-            how="left",
-            coalesce=True,
+        .join(tm_trips, on="TRIP_ID", how="left")
+        .join(tm_geo_nodes, on="GEO_NODE_ID", how="left")
+        .join(tm_time_points, on="TIME_POINT_ID", how="left")
+        .join(tm_routes, on="ROUTE_ID", how="left")
+        .with_columns(
+            (
+                pl.col("CALENDAR_ID").cast(pl.String).str.to_date("1%Y%m%d").alias("service_date"),
+                pl.col("TRIP_SERIAL_NUMBER").cast(pl.String).alias("trip_id"),
+                pl.col("GEO_NODE_ABBR").alias("stop_id"),
+                pl.col("PATTERN_GEO_NODE_SEQ").cast(pl.Int64).alias("tm_stop_sequence"),
+                pl.col("TIME_POINT_ID").cast(pl.Int64).alias("timepoint_id"),
+                pl.col("TIME_POINT_ABBR").cast(pl.String).alias("timepoint_abbr"),
+                pl.col("TIME_PT_NAME").cast(pl.String).alias("timepoint_name"),
+                pl.col("PATTERN_ID").cast(pl.Int64).alias("pattern_id"),
+                (
+                    pl.col("CALENDAR_ID").cast(pl.String).str.to_datetime("1%Y%m%d", time_unit="us")
+                    + pl.duration(seconds=pl.col("SCHEDULED_TIME"))
+                ).alias("plan_stop_departure_dt"),
+                pl.col("SCHEDULED_TIME").cast(pl.Int64).alias("tm_scheduled_time_sam"),
+                pl.col(["PATTERN_GEO_NODE_SEQ"]).rank(method="dense").over(["PATTERN_ID"]).alias("timepoint_order"),
+                pl.col("ROUTE_ABBR").str.strip_chars_start(pl.lit("0")).alias("route_id"),
+            )
         )
-        .join(tm_geo_nodes, on="GEO_NODE_ID", how="left", coalesce=True)
-        .join(tm_time_points, on="TIME_POINT_ID", how="left", coalesce=True)
     )
     # TRIP_ID or [TRIP_SERIAL_NUMBER, PATTERN_ID] uniquely identify a TM "Trip".
     # TRIP_SERIAL_NUMBER is the publicly facing number (and gets aliased to "TRIP_ID" below)
-    tm_sequences = tm_trip_geo_tp.group_by(["TRIP_ID"]).agg(
+    tm_sequences = stop_crossings.group_by(["TRIP_ID"]).agg(
         pl.col("PATTERN_GEO_NODE_SEQ").max().alias("tm_planned_sequence_end"),
         pl.col("PATTERN_GEO_NODE_SEQ").min().alias("tm_planned_sequence_start"),
     )
 
-    tm_schedule = tm_trip_geo_tp_full.with_columns(
-        pl.col("TRIP_SERIAL_NUMBER").cast(pl.String).alias("trip_id"), pl.col("GEO_NODE_ABBR").alias("stop_id")
-    ).join(
-        tm_sequences,
-        on="TRIP_ID",
-        how="left",
-        coalesce=True,
+    tm_schedule = (
+        stop_crossings.join(
+            tm_sequences,
+            on="TRIP_ID",
+            how="left",
+            coalesce=True,
+        )
+        .sort("VEHICLE_ID", nulls_last=True)
+        .unique(
+            subset=[
+                "trip_id",
+                "route_id",
+                "tm_stop_sequence",
+                "plan_stop_departure_dt",
+            ],
+            keep="first",  # if there are ties, keep the record with the non-null vehicle ID
+            maintain_order=True,
+        )
```

Comment:
does this need to happen? can we just join on TRIP_GEO_NODE_PATTERN_SEQ? if not, comment liberally